docs/zic.8.rst
ZIC
===

| `NAME <#NAME>`__
| `SYNOPSIS <#SYNOPSIS>`__
| `DESCRIPTION <#DESCRIPTION>`__
| `EXTENDED EXAMPLE <#EXTENDED%20EXAMPLE>`__
| `NOTES <#NOTES>`__
| `FILE <#FILE>`__
| `SEE ALSO <#SEE%20ALSO>`__

--------------

` <>`__

NAME
----

zic − time zone compiler

` <>`__

SYNOPSIS
--------

**zic** [ *option* ... ] [ *filename* ... ]

` <>`__

DESCRIPTION
-----------

*Zic* reads text from the file(s) named on the command line and creates
the time conversion information files specified in this input. If a
*filename* is “\ ``-``\ ”, the standard input is read.

These options are available: `` --``\ **version**

Output version information and exit.

``-``\ **d** *directory*

Create time conversion information files in the named directory rather
than in the standard directory named below.

``-``\ **l** *timezone*

Use the given time zone as local time. *Zic* will act as if the input
contained a link line of the form

Link

*timezone*

localtime

``-``\ **p** *timezone*

Use the given time zone’s rules when handling POSIX-format time zone
environment variables. *Zic* will act as if the input contained a link
line of the form

Link

*timezone*

posixrules

``-``\ **L** *leapsecondfilename*

Read leap second information from the file with the given name. If this
option is not used, no leap second information appears in output files.

+--------------------+--------------------+--------------------+--------------------+
|                    | ``-``\ **v**       |                    | Be more verbose,   |
|                    |                    |                    | and complain about |
|                    |                    |                    | the following      |
|                    |                    |                    | situations:        |
+--------------------+--------------------+--------------------+--------------------+

The input specifies a link to a link.

A year that appears in a data file is outside the range of years
representable by *time*\ (2) values.

A time of 24:00 or more appears in the input. Pre-1998 versions of *zic*
prohibit 24:00, and pre-2007 versions prohibit times greater than 24:00.

A rule goes past the start or end of the month. Pre-2004 versions of
*zic* prohibit this.

The output file does not contain all the information about the long-term
future of a zone, because the future cannot be summarized as an extended
POSIX TZ string. For example, as of 2013 this problem occurs for Iran’s
daylight-saving rules for the predicted future, as these rules are based
on the Iranian calendar, which cannot be represented.

The output contains data that may not be handled properly by client code
designed for older *zic* output formats. These compatibility issues
affect only time stamps before 1970 or after the start of 2038.

A time zone abbreviation has fewer than 3 characters. POSIX requires at
least 3.

An output file name contains a byte that is not an ASCII letter,
“\ ``-``\ ”, “/”, or “\_”; or it contains a file name component that
contains more than 14 bytes or that starts with “\ ``-``\ ”.

+--------------------+--------------------+--------------------+--------------------+
|                    | ``-``\ **s**       |                    | Limit time values  |
|                    |                    |                    | stored in output   |
|                    |                    |                    | files to values    |
|                    |                    |                    | that are the same  |
|                    |                    |                    | whether they’re    |
|                    |                    |                    | taken to be signed |
|                    |                    |                    | or unsigned. You   |
|                    |                    |                    | can use this       |
|                    |                    |                    | option to generate |
|                    |                    |                    | SVVS-compatible    |
|                    |                    |                    | files.             |
+--------------------+--------------------+--------------------+--------------------+

Input files should be text files, that is, they should be a series of
zero or more lines, each ending in a newline byte and containing at most
511 bytes, and without any NUL bytes. The input text’s encoding is
typically UTF-8 or ASCII; it should have a unibyte representation for
the POSIX Portable Character Set (PPCS) ⟨http://pubs.opengroup.org/
onlinepubs/9699919799/basedefs/V1\_chap06.html⟩ and the encoding’s
non-unibyte characters should consist entirely of non-PPCS bytes.
Non-PPCS characters typically occur only in comments: although output
file names and time zone abbreviations can contain nearly any character,
other software will work better if these are limited to the restricted
syntax described under the ``-``\ **v** option.

Input lines are made up of fields. Fields are separated from one another
by one or more white space characters. The white space characters are
space, form feed, carriage return, newline, tab, and vertical tab.
Leading and trailing white space on input lines is ignored. An unquoted
sharp character (#) in the input introduces a comment which extends to
the end of the line the sharp character appears on. White space
characters and sharp characters may be enclosed in double quotes (") if
they’re to be used as part of a field. Any line that is blank (after
comment stripping) is ignored. Non-blank lines are expected to be of one
of three types: rule lines, zone lines, and link lines.

Names (such as month names) must be in English and are case insensitive.
Abbreviations, if used, must be unambiguous in context.

A rule line has the form

Rule NAME FROM TO TYPE IN ON AT SAVE LETTER/S

For example:

Rule US 1967 1973 ``-`` Apr lastSun 2:00 1:00 D

The fields that make up a rule line are:

+--------------------+--------------------+--------------------+--------------------+
|                    | **NAME**           |                    | Gives the          |
|                    |                    |                    | (arbitrary) name   |
|                    |                    |                    | of the set of      |
|                    |                    |                    | rules this rule is |
|                    |                    |                    | part of.           |
+--------------------+--------------------+--------------------+--------------------+
|                    | **FROM**           |                    | Gives the first    |
|                    |                    |                    | year in which the  |
|                    |                    |                    | rule applies. Any  |
|                    |                    |                    | integer year can   |
|                    |                    |                    | be supplied; the   |
|                    |                    |                    | proleptic          |
|                    |                    |                    | Gregorian calendar |
|                    |                    |                    | is assumed. The    |
|                    |                    |                    | word **minimum**   |
|                    |                    |                    | (or an             |
|                    |                    |                    | abbreviation)      |
|                    |                    |                    | means the minimum  |
|                    |                    |                    | year representable |
|                    |                    |                    | as an integer. The |
|                    |                    |                    | word **maximum**   |
|                    |                    |                    | (or an             |
|                    |                    |                    | abbreviation)      |
|                    |                    |                    | means the maximum  |
|                    |                    |                    | year representable |
|                    |                    |                    | as an integer.     |
|                    |                    |                    | Rules can describe |
|                    |                    |                    | times that are not |
|                    |                    |                    | representable as   |
|                    |                    |                    | time values, with  |
|                    |                    |                    | the                |
|                    |                    |                    | unrepresentable    |
|                    |                    |                    | times ignored;     |
|                    |                    |                    | this allows rules  |
|                    |                    |                    | to be portable     |
|                    |                    |                    | among hosts with   |
|                    |                    |                    | differing time     |
|                    |                    |                    | value types.       |
+--------------------+--------------------+--------------------+--------------------+
|                    | **TO**             |                    | Gives the final    |
|                    |                    |                    | year in which the  |
|                    |                    |                    | rule applies. In   |
|                    |                    |                    | addition to        |
|                    |                    |                    | **minimum** and    |
|                    |                    |                    | **maximum** (as    |
|                    |                    |                    | above), the word   |
|                    |                    |                    | **only** (or an    |
|                    |                    |                    | abbreviation) may  |
|                    |                    |                    | be used to repeat  |
|                    |                    |                    | the value of the   |
|                    |                    |                    | **FROM** field.    |
+--------------------+--------------------+--------------------+--------------------+
|                    | **TYPE**           |                    | should be          |
|                    |                    |                    | “\ ``-``\ ” and is |
|                    |                    |                    | present for        |
|                    |                    |                    | compatibility with |
|                    |                    |                    | older versions of  |
|                    |                    |                    | *zic* in which it  |
|                    |                    |                    | could contain year |
|                    |                    |                    | types.             |
+--------------------+--------------------+--------------------+--------------------+
|                    | **IN**             |                    | Names the month in |
|                    |                    |                    | which the rule     |
|                    |                    |                    | takes effect.      |
|                    |                    |                    | Month names may be |
|                    |                    |                    | abbreviated.       |
+--------------------+--------------------+--------------------+--------------------+
|                    | **ON**             |                    | Gives the day on   |
|                    |                    |                    | which the rule     |
|                    |                    |                    | takes effect.      |
|                    |                    |                    | Recognized forms   |
|                    |                    |                    | include:           |
+--------------------+--------------------+--------------------+--------------------+

+--------------------------+--------------------------+--------------------------+
|                          |                          | 5                        |
+--------------------------+--------------------------+--------------------------+
|                          |                          | the fifth of the month   |
+--------------------------+--------------------------+--------------------------+
|                          |                          | lastSun                  |
+--------------------------+--------------------------+--------------------------+
|                          |                          | the last Sunday in the   |
|                          |                          | month                    |
+--------------------------+--------------------------+--------------------------+
|                          |                          | lastMon                  |
+--------------------------+--------------------------+--------------------------+
|                          |                          | the last Monday in the   |
|                          |                          | month                    |
+--------------------------+--------------------------+--------------------------+
|                          |                          | Sun>=8                   |
+--------------------------+--------------------------+--------------------------+
|                          |                          | first Sunday on or after |
|                          |                          | the eighth               |
+--------------------------+--------------------------+--------------------------+
|                          |                          | Sun<=25                  |
+--------------------------+--------------------------+--------------------------+
|                          |                          | last Sunday on or before |
|                          |                          | the 25th                 |
+--------------------------+--------------------------+--------------------------+

Names of days of the week may be abbreviated or spelled out in full.
Note that there must be no spaces within the **ON** field.

+--------------------+--------------------+--------------------+--------------------+
|                    | **AT**             |                    | Gives the time of  |
|                    |                    |                    | day at which the   |
|                    |                    |                    | rule takes effect. |
|                    |                    |                    | Recognized forms   |
|                    |                    |                    | include:           |
+--------------------+--------------------+--------------------+--------------------+

+--------------------------+--------------------------+--------------------------+
|                          |                          | 2                        |
+--------------------------+--------------------------+--------------------------+
|                          |                          | time in hours            |
+--------------------------+--------------------------+--------------------------+
|                          |                          | 2:00                     |
+--------------------------+--------------------------+--------------------------+
|                          |                          | time in hours and        |
|                          |                          | minutes                  |
+--------------------------+--------------------------+--------------------------+
|                          |                          | 15:00                    |
+--------------------------+--------------------------+--------------------------+
|                          |                          | 24-hour format time (for |
|                          |                          | times after noon)        |
+--------------------------+--------------------------+--------------------------+
|                          |                          | 1:28:14                  |
+--------------------------+--------------------------+--------------------------+
|                          |                          | time in hours, minutes,  |
|                          |                          | and seconds              |
+--------------------------+--------------------------+--------------------------+
|                          |                          | ``-``                    |
+--------------------------+--------------------------+--------------------------+
|                          |                          | equivalent to 0          |
+--------------------------+--------------------------+--------------------------+

where hour 0 is midnight at the start of the day, and hour 24 is
midnight at the end of the day. Any of these forms may be followed by
the letter **w** if the given time is local “wall clock” time, **s** if
the given time is local “standard” time, or **u** (or **g** or **z**) if
the given time is universal time; in the absence of an indicator, wall
clock time is assumed. The intent is that a rule line describes the
instants when a clock/calendar set to the type of time specified in the
**AT** field would show the specified date and time of day.

+--------------------+--------------------+--------------------+--------------------+
|                    | **SAVE**           |                    | Gives the amount   |
|                    |                    |                    | of time to be      |
|                    |                    |                    | added to local     |
|                    |                    |                    | standard time when |
|                    |                    |                    | the rule is in     |
|                    |                    |                    | effect. This field |
|                    |                    |                    | has the same       |
|                    |                    |                    | format as the      |
|                    |                    |                    | **AT** field       |
|                    |                    |                    | (although, of      |
|                    |                    |                    | course, the **w**  |
|                    |                    |                    | and **s** suffixes |
|                    |                    |                    | are not used).     |
|                    |                    |                    | Only the sum of    |
|                    |                    |                    | standard time and  |
|                    |                    |                    | this amount        |
|                    |                    |                    | matters; for       |
|                    |                    |                    | example, *zic*     |
|                    |                    |                    | does not           |
|                    |                    |                    | distinguish a      |
|                    |                    |                    | 10:30 standard     |
|                    |                    |                    | time plus an 0:30  |
|                    |                    |                    | **SAVE** from a    |
|                    |                    |                    | 10:00 standard     |
|                    |                    |                    | time plus a 1:00   |
|                    |                    |                    | **SAVE**.          |
+--------------------+--------------------+--------------------+--------------------+

**LETTER/S**

Gives the “variable part” (for example, the “S” or “D” in “EST” or
“EDT”) of time zone abbreviations to be used when this rule is in
effect. If this field is “\ ``-``\ ”, the variable part is null.

A zone line has the form

Zone

NAME

GMTOFF

RULES/SAVE

FORMAT

[UNTILYEAR [MONTH [DAY [TIME]]]]

For example:

Zone

Australia/Adelaide

9:30

Aus

AC%sT

1971 Oct 31 2:00

The fields that make up a zone line are:

+--------------------+--------------------+--------------------+--------------------+
|                    | **NAME**           |                    | The name of the    |
|                    |                    |                    | time zone. This is |
|                    |                    |                    | the name used in   |
|                    |                    |                    | creating the time  |
|                    |                    |                    | conversion         |
|                    |                    |                    | information file   |
|                    |                    |                    | for the zone. It   |
|                    |                    |                    | should not contain |
|                    |                    |                    | a file name        |
|                    |                    |                    | component “.” or   |
|                    |                    |                    | “..”; a file name  |
|                    |                    |                    | component is a     |
|                    |                    |                    | maximal substring  |
|                    |                    |                    | that does not      |
|                    |                    |                    | contain “/”.       |
+--------------------+--------------------+--------------------+--------------------+

**GMTOFF**

The amount of time to add to UT to get standard time in this zone. This
field has the same format as the **AT** and **SAVE** fields of rule
lines; begin the field with a minus sign if time must be subtracted from
UT.

**RULES/SAVE**

The name of the rule(s) that apply in the time zone or, alternately, an
amount of time to add to local standard time. If this field is ``-``
then standard time always applies in the time zone. When an amount of
time is given, only the sum of standard time and this amount matters.

**FORMAT**

The format for time zone abbreviations in this time zone. The pair of
characters **%s** is used to show where the “variable part” of the time
zone abbreviation goes. Alternately, a format can use the pair of
characters **%z** to stand for the UTC offset in the form ±\ *hh*,
±\ *hhmm*, or ±\ *hhmmss*, using the shortest form that does not lose
information, where *hh*, *mm*, and *ss* are the hours, minutes, and
seconds east (+) or west (−) of UTC. Alternately, a slash (/) separates
standard and daylight abbreviations. To conform to POSIX, a time zone
abbreviation should contain only alphanumeric ASCII characters, "+" and
"``-``".

**UNTILYEAR [MONTH [DAY [TIME]]]**

The time at which the UT offset or the rule(s) change for a location. It
is specified as a year, a month, a day, and a time of day. If this is
specified, the time zone information is generated from the given UT
offset and rule change until the time specified, which is interpreted
using the rules in effect just before the transition. The month, day,
and time of day have the same format as the IN, ON, and AT fields of a
rule; trailing fields can be omitted, and default to the earliest
possible value for the missing fields.

The next line must be a “continuation” line; this has the same form as a
zone line except that the string “Zone” and the name are omitted, as the
continuation line will place information starting at the time specified
as the “until” information in the previous line in the file used by the
previous line. Continuation lines may contain “until” information, just
as zone lines do, indicating that the next line is a further
continuation.

If a zone changes at the same instant that a rule would otherwise take
effect in the earlier zone or continuation line, the rule is ignored. In
a single zone it is an error if two rules take effect at the same
instant, or if two zone changes take effect at the same instant.

A link line has the form

Link

TARGET

LINK-NAME

For example:

Link

Europe/Istanbul

Asia/Istanbul

The **TARGET** field should appear as the **NAME** field in some zone
line. The **LINK-NAME** field is used as an alternate name for that
zone; it has the same syntax as a zone line’s **NAME** field.

Except for continuation lines, lines may appear in any order in the
input. However, the behavior is unspecified if multiple zone or link
lines define the same name, or if the source of one link line is the
target of another.

Lines in the file that describes leap seconds have the following form:

Leap

YEAR

MONTH

DAY

HH:MM:SS

CORR

R/S

For example:

Leap

1974

Dec

31

23:59:60

+

S

The **YEAR**, **MONTH**, **DAY**, and **HH:MM:SS** fields tell when the
leap second happened. The **CORR** field should be “+” if a second was
added or “\ ``-``\ ” if a second was skipped. The **R/S** field should
be (an abbreviation of) “Stationary” if the leap second time given by
the other fields should be interpreted as UTC or (an abbreviation of)
“Rolling” if the leap second time given by the other fields should be
interpreted as local wall clock time.

` <>`__

EXTENDED EXAMPLE
----------------

Here is an extended example of *zic* input, intended to illustrate many
of its features.

+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | # Rule | NAME   | FROM   | TO     | TYPE   | IN     | ON     | AT     | SAVE   |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | LETTER |
|        |        |        |        |        |        |        |        |        |        | /S     |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | Swiss  | 1941   | 1942   | ``-``  | May    | Mon>=1 | 1:00   | 1:00   |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | S      |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | Swiss  | 1941   | 1942   | ``-``  | Oct    | Mon>=1 | 2:00   | 0      |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | ``-``  |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | EU     | 1977   | 1980   | ``-``  | Apr    | Sun>=1 | 1:00u  | 1:00   |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | S      |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | EU     | 1977   | only   | ``-``  | Sep    | lastSu | 1:00u  | 0      |
|        |        |        |        |        |        |        |        | n      |        |        |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | ``-``  |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | EU     | 1978   | only   | ``-``  | Oct    | 1      | 1:00u  | 0      |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | ``-``  |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | EU     | 1979   | 1995   | ``-``  | Sep    | lastSu | 1:00u  | 0      |
|        |        |        |        |        |        |        |        | n      |        |        |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | ``-``  |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | EU     | 1981   | max    | ``-``  | Mar    | lastSu | 1:00u  | 1:00   |
|        |        |        |        |        |        |        |        | n      |        |        |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | S      |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        | Rule   | EU     | 1996   | max    | ``-``  | Oct    | lastSu | 1:00u  | 0      |
|        |        |        |        |        |        |        |        | n      |        |        |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+
|        |        |        |        |        |        |        |        |        |        | ``-``  |
+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+--------+

+------------+------------+------------+------------+------------+------------+------------+
|            |            | # Zone     |            |            |            |            |
+------------+------------+------------+------------+------------+------------+------------+
|            |            | NAME       |            | GMTOFF     |            |            |
+------------+------------+------------+------------+------------+------------+------------+
|            |            |            |            | RULES/SAVE |            | FORMAT     |
+------------+------------+------------+------------+------------+------------+------------+
|            |            |            |            |            |            | UNTIL      |

tzdata/zic.py
import textwrap

import sys

from datetime import datetime

HEADER = """\
from tzdata.classes import *
from datetime import *


"""


def lines(input):
    """Remove comments and empty lines"""
    for raw_line in input:
        line = raw_line.strip()
        if line and not line.startswith('#'):
            yield strip_comments(line)


def strip_comments(line):
    quoted = False
    for i, c in enumerate(line):
        if c == '"':
            quoted = not quoted
        elif c == "#" and not quoted:
            return line[:i].strip()
    return line


OBSERVANCE_TEMPLATE = """\
Observance(gmtoff={},
           rules={},
           format='{}',
           until={}),
"""


def compile(infile, outfile=None):
    with open(infile) as input:
        if outfile is None:
            compile_stream(input, sys.stdout)
        else:
            with open(outfile, 'w') as output:
                compile_stream(input, output)


def compile_stream(input, output, header=HEADER):
    output.write(header)
    observances = state = None
    zones = {}
    rules = {}
    for line in lines(input):
        fields = line.split()
        if fields[0] == 'Zone':
            names = fields[1].split('/')
            z = zones
            for name in names:
                z = z.setdefault(name, {})
            observances = z.setdefault('observances', [])
            state = 'Zone'
            del fields[:2]
        elif fields[0] == 'Rule':
            rules.setdefault(fields[1], []).append(fields[2:])
        if state == 'Zone':
            gmtoff, zone_rules, format = fields[:3]
            until = format_until(fields[3:])
            if until is None:
                state = None

            observances.append(
                format_observance(gmtoff, zone_rules, format, until))
    print_rules(rules, file=output)
    print_zones(zones, file=output)

RULE_TEMPLATE = ('Rule({}, {}, {}, {}, {},\n'
                 '     at={},\n'
                 '     save={}, letters={!r})')


def format_rule(begin, end, type, in_month, on, at, save, letters):
    begin = int(begin)
    if end == 'only':
        end = begin + 1
    elif end == 'max':
        end = 10000
    else:
        end = int(end) + 1
    if type == '-':
        type = None
    if letters == '-':
        letters = ''
    at = format_at(at)
    save = format_time(save)
    return RULE_TEMPLATE.format(begin, end, type, in_month,
                                on, at, save, letters)

TIME_FORMATS = ['%H', '%H:%M', "%H:%M:%S"]
TIME_TYPES = {
    'w': 'wall',
    'u': 'utc',
    'g': 'utc',
    'z': 'utc',
    's': 'std',
}


def format_time(t):
    if t == '-':
        return 'timedelta(0)'
    if t.startswith('24'):
        return 'timedelta(1)'
    n = t.count(':')
    fmt = TIME_FORMATS[n]
    t = datetime.strptime(t, fmt).time()
    args = ['hours={0.hour}', 'minutes={0.minute}', 'seconds={0.second)']
    template = 'timedelta(%s)' % ', '.join(args[:n+1])
    return template.format(t)


def format_at(at):
    try:
        time_type = TIME_TYPES[at[-1]]
    except KeyError:
        time_type = 'wall'
    else:
        at = at[:-1]
    return '(%s, %r)' % (format_time(at), time_type)


def print_rules(rules, file):
    prefix = ' ' * 8
    for name, lines in rules.items():
        file.write('class %s(Rules):\n'
                   '    name ="%s"\n'
                   '    rules = [\n' % (rules_name(name), name))
        for args in lines:
            rule = format_rule(*args)
            file.write(textwrap.indent(rule, prefix) + ',\n')
        file.write('    ]\n\n')

TIME_UNITS = 'hours', 'minutes', 'seconds'


def format_until(until):
    n = len(until)
    if n == 0:
        return None
    if n == 1:
        return int(until[0])
    return '(%s)' % ', '.join(repr(u) for u in until)


def format_delta(delta):
    sign = ''
    if delta.startswith('-'):
        sign = '-'
        delta = delta[1:]
    args = ['%s=%s' % (unit, int(value))
            for unit, value in zip(TIME_UNITS, delta.split(':'))]
    return '%stimedelta(%s)' % (sign, ', '.join(args))


def format_observance(gmtoff, rules, format, until):
    if rules == '-':
        rules = None
    elif ':' in rules:
        rules = "'%s'" % rules
    else:
        rules = rules_name(rules)
    return OBSERVANCE_TEMPLATE.format(format_delta(gmtoff),
                                      rules, format, until)


def print_zones(zones, file, indent=0):
    for name, info in sorted(zones.items()):
        try:
            observances = info['observances']
        except KeyError:
            file.write(indent * ' ' + 'class %s:\n' % name)
            print_zones(info, file, indent + 4)
        else:
            prefix = indent * ' '
            file.write(prefix + 'class %s(Zone):\n' % zone_name(name))
            file.write(prefix + '    name = %r\n' % name)
            file.write(prefix + '    observances = [\n')
            for observance in observances:
                file.write(textwrap.indent(observance, prefix + 8 * ' '))
            file.write(prefix + '%s]\n' % (4 * ' '))


def rules_name(name):
    return name.replace('-', '_')

zone_name = rules_name

if __name__ == '__main__':
    if len(sys.argv) < 2:
        print("Usage: zic infile [outfile]")
        sys.exit(1)
    compile(*sys.argv[1:])
